helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java

```diff
@@ -242,16 +230,16 @@ public void updatePreviousAssignedTasksStatus(
         }
           break;
         case COMPLETED: {
-          // The task has completed on this partition. Mark as such in the context object.
-          donePartitions.add(pId);
+          // The task has completed on this partition. Drop it from the instance and add it to assignedPartitions in
+          // order to avoid scheduling it again in this pipeline.
+          assignedPartitions.get(instance).add(pId);
+          paMap.put(pId, new PartitionAssignment(instance, TaskPartitionState.DROPPED.name()));
```

Comment:
I don't know if this would be the correct thing to do.

Sure, once a task is complete, it could be dropped from the node, but you also want to keep the state of the task in COMPLETED in JobContext.

I'm afraid doing this might cause the otherwise-completed task's state to change to DROPPED. Then we would have no way of knowing if this task just got dropped or completed and dropped.

Reply:
You are right. I also had this concern. But note that we do not change the JobContext to DROPPED. Let’s say controller sends COMPLETED to DROPPED message to the participant. Now, the current state will be COMPLETED until participant processes the message. Once it is processed, the current state will be null, and that task will not be considered again. So, to answer your question, it is totally depending on the behavior of the participant. If participant set the current state  to DROPPED, which is not the case here, your concern was correct. Since participant removes the current state of the task, we do not need to be worry about that case.